disco/cli/make_upgrade_tables.py

```diff
@@ -0,0 +1,150 @@
+
+import csv
+import itertools
+import logging
+from collections import namedtuple
+from concurrent.futures import ProcessPoolExecutor
+from pathlib import Path
+
+import click
+import pandas as pd
+
+from jade.common import CONFIG_FILE, JOBS_OUTPUT_DIR
+from jade.loggers import setup_logging
+from jade.jobs.job_configuration_factory import create_config_from_file
+from jade.jobs.results_aggregator import ResultsAggregator
+
+from disco.pipelines.utils import ensure_jade_pipeline_output_dir
+
+logger = logging.getLogger(__name__)
+
+JobInfo = namedtuple(
+    "JobInfo", ["name", "substation", "feeder", "placement", "sample", "penetration_level"]
+)
+
+
+def parse_batch_results(output_dir):
+    """Parse the results from all jobs in a JADE output directory."""
+    output_path = Path(output_dir)
+    config = create_config_from_file(output_path / CONFIG_FILE)
+    jobs = []
+    results = ResultsAggregator.list_results(output_dir)
+    result_lookup = {x.name: x for x in results}
+    for job in config.iter_jobs():
+        if job.name not in result_lookup:
+            logger.info("Skip missing job %s", job.name)
+            continue
+        if result_lookup[job.name].return_code != 0:
+            logger.info("Skip failed job %s", job.name)
+            continue
+        jobs.append(job)
+
+    upgrade_summary_table = []
+    total_upgrade_costs_table = []
+
+    with ProcessPoolExecutor() as executor:
+        for result in executor.map(parse_job_results, jobs, itertools.repeat(output_path)):
+            upgrade_summary_table += result[0]
+            total_upgrade_costs_table += result[1]
+
+    serialize_table(upgrade_summary_table, output_path / "upgrade_summary.csv")
+    serialize_table(total_upgrade_costs_table, output_path / "total_upgrade_costs.csv")
+
+
+def parse_job_results(job, output_path):
+    job_path = output_path / JOBS_OUTPUT_DIR / job.name
+    deployment = job.model.deployment
+    job_info = JobInfo(
+        name=job.name,
+        substation=deployment.substation,
+        feeder=deployment.feeder,
+        placement=deployment.project_data["placement"],
+        sample=deployment.project_data["sample"],
+        penetration_level=deployment.project_data["penetration_level"],
+    )
+    upgrade_summary_table = get_upgrade_summary_table(job_path, job_info)
+    total_upgrade_costs_table = get_total_upgrade_costs_table(job_path, job_info)
+    
+    return (
+        upgrade_summary_table,
+        total_upgrade_costs_table
+    )
+
+
+def get_upgrade_summary_table(job_path, job_info):
+
+    def _get_upgrade_summary(upgrade_type, upgrade_summary_file):
+        if not upgrade_summary_file.exists():
+            return []
+        
+        try:
+            df = pd.read_csv(upgrade_summary_file)
+        except pd.errors.EmptyDataError as e:
+            logger.error("Failed to parse upgrade summary file - '%s'", upgrade_summary_file)
+            logger.exception(e)
```

Comment:
I think you can consolidate these. And no need to log `e`. `logger.exception` will log your message and then append the backtrace.

Reply:
you are right, consolidated.